### PR TITLE
Run container app as a service

### PR DESCRIPTION
**Pull request by giggsoff**
Ability to run container app as a service. If Service flag set
in config and app has "org.mobyproject.config" label we will configure
the app accordingly (adjust mount, capabilities, namespaces...).

### CODE REVIEW

File: pkg/pillar/types/domainmgrtypes.go [hunk 1]
```diff
@@ -43,6 +43,11 @@ type DomainConfig struct {
 
 	// MetaDataType for select type of metadata service for app
 	MetaDataType MetaDataType
+
+	// Privileged indicates that we want to run app as a service
+	Privileged bool
```
Comment (eriknordmark) [inline pkg/pillar/types/domainmgrtypes.go:48]: Isn't this name a bit misleading? IsSystemService or IsServiceContainer seems more clear to me.
For instance the edgeview container does not have all priviledges.

**Review by eriknordmark — COMMENTED** [review]
The mechanism here is interesting, and I think we should have an email discussion on the lf-edge EVE mailing list about the problems we want to solve.

In my mind being able to lift the debug and edgeview containers out from the image and have them be downloadable and runnable as IsServiceContainer is a problem I'd like to solve.
I think that means they would be "prepared" since they are built by linuxkit.

The replacement logic seems like overkill; folks would need to know how to use it and it only matters when there are multiple conflicting system services; how would the user know this might be the case? (Unless the idea is to be able to replace e.g., the debug container in the image with a downloaded debug container.)

I think we should make the IsSystemService boolean (and if we need a separate Priviledged bool) part of the EVE API so that the controller is well aware (and can apply policy checks) to any such containers.

But @petr-zededa can you start a thread on the EVE mailing list about this?

**Comment by deitch** [comment]
I want to be able to contribute to the discussion, but I don't quite get the problem being solved, or how it is being solved. I would favour that thread on EVE mailing list, but please start with the high-level: here is the problem, here is the design for it without too much detail, i.e. what things happen now, what would change. The specific flags or code change are less relevant.

@eriknordmark does this belong in a mailing list proposal, or should it be a design proposal on the wiki?

**Comment by zedi-pramodh** [comment]
I agree with @deitch. Looks like a huge change with no design documentation. Please help us understand.

**Comment by giggsoff** [comment]
I submitted a proposal: https://lists.lfedge.org/g/eve-tsc/message/913

File: api/proto/config/appconfig.proto [hunk 2]
```diff
@@ -126,6 +126,10 @@ message AppInstanceConfig {
   // Non-Zero value -> After EVE is ready to start application instance, wait for the
   // given amount of time before starting the respective application instance.
   uint32 start_delay_in_seconds = 19;
+
+  // config_mode - define what to use as a base configuration
+  // for environment of app instance
+  ConfigMode config_mode = 20;
```
Comment (rvs) [inline api/proto/config/appconfig.proto:132]: I've got questions about this part -- see the email thread @giggsoff

File: api/proto/config/appconfig.proto [hunk 3]
```diff
@@ -126,6 +126,11 @@ message AppInstanceConfig {
   // Non-Zero value -> After EVE is ready to start application instance, wait for the
   // given amount of time before starting the respective application instance.
   uint32 start_delay_in_seconds = 19;
+
+  // service - run app instance as a service
```
Comment (eriknordmark) [inline api/proto/config/appconfig.proto:130]: I'd suggest "as a service container in the host/dom0"

File: docs/ECOS.md [hunk 4]
```diff
@@ -205,6 +205,15 @@ Please note, that change of environment variables require restart of the applica
 After preparation done we chroot into /mnt and run cmd from cmdline under specified user and group, output goes to
 /dev/console and is accessible from log of ECO.
 
+## Ability to override default capabilities, mounts and namespaces
+
+By default, we bootstrap OCI configuration using [oci.WithDefaultSpec](https://pkg.go.dev/github.com/containerd/containerd/oci#WithDefaultSpec) function. It fit our expectations for generic ECO we want to run on EVE-OS to be isolated and has limited privileges.
```
Comment (eriknordmark) [inline docs/ECOS.md:210]: Do we use this with NOHYPER? If not we should disable that and only allow NOHYPER when service=true.

File: pkg/pillar/go.mod [hunk 5]
```diff
@@ -12,9 +12,7 @@ require (
 	github.com/containerd/typeurl v1.0.2
 	github.com/cshari-zededa/eve-tpm2-tools v0.0.4
 	github.com/digitalocean/go-qemu v0.0.0-20181112162955-dd7bb9c771b8
-	github.com/docker/cli v20.10.13+incompatible // indirect
-	github.com/docker/distribution v2.8.1+incompatible // indirect
-	github.com/docker/docker v20.10.13+incompatible
+	github.com/docker/docker v20.10.17+incompatible
```
Comment (eriknordmark) [inline pkg/pillar/go.mod:15]: Any particular reason for updating these vendored packages as part of this PR?
If we need to update vendored packages it presumably makes sense to do that as part of a separate PR with only those changes.

Comment (giggsoff) [reply pkg/pillar/go.mod:15]: They come with import of `github.com/linuxkit/linuxkit/src/cmd/linuxkit` (as we use `github.com/linuxkit/linuxkit/src/cmd/linuxkit/mobi` to extract data from label).

File: api/proto/config/appconfig.proto (hunk 3 above)
Comment (giggsoff) [reply api/proto/config/appconfig.proto]: Done, thank you

File: docs/ECOS.md (hunk 4 above)
Comment (giggsoff) [reply docs/ECOS.md]: I put few comments about that in doc. Correct, we use WithDefaultSpec if there is no hypervisor available on the system (or if we enforce to run as container using NOHYPER VmMode).

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 6]
```diff
@@ -632,6 +632,11 @@ func parseAppInstanceConfig(getconfigCtx *getconfigContext,
 		appInstance.FixedResources.DisableLogs = cfgApp.Fixedresources.DisableLogs
 		appInstance.MetaDataType = types.MetaDataType(cfgApp.MetaDataType)
 		appInstance.Delay = time.Duration(cfgApp.StartDelayInSeconds) * time.Second
+		appInstance.Service = cfgApp.Service
+		if appInstance.Service && appInstance.FixedResources.VirtualizationMode != types.NOHYPER {
+			log.Warnf("MobyLabel ConfigMode app instance %s should have NOHYPER VirtualizationMode", appInstance.UUIDandVersion.UUID)
+			appInstance.FixedResources.VirtualizationMode = types.NOHYPER
+		}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go:639]: Can we make this return an error instead of just logging a warning? That would be easier if the check is moved from zedagent to zedmanager or domainmgr unless parseAppInstanceConfig() already has the structure to add parse errors to appInstanceConfig (which zedmanager turns into errors in AppInstanceStatus). (We do have such a structure for some other objects in parseconfig.go but I'm not sure about AppInstanceConfig.)

Comment (giggsoff) [reply pkg/pillar/cmd/zedagent/parseconfig.go]: Make sense, thank you! And we have `checkAndPublishAppInstanceConfig` which is perfectly fit.

**Review by eriknordmark — APPROVED** [review]
LGTM

**Review by eriknordmark — APPROVED** [review]
Run eden again
